src/chainstate/stacks/boot/exit-at-rc.clar

```diff
@@ -0,0 +1,199 @@
+;; The .exit-at-rc contract
+;; Error codes
+(define-constant ERR_AMOUNT_NOT_POSITIVE 2)
+(define-constant ERR_PREVIOUS_VOTE_VALID 7)
+(define-constant ERR_ALREADY_VETOED 9)
+(define-constant ERR_UNAUTHORIZED_CALLER 10)
+(define-constant ERR_VOTER_NOT_STACKING 13)
+(define-constant ERR_BURN_BLOCK_HEIGHT_TOO_LOW 14)
+(define-constant ERR_FETCHING_BLOCK_INFO 16)
+(define-constant ERR_NOT_ALLOWED 19)
+(define-constant ERR_INVALID_PROPOSED_RC 21)
+
+;; Constants
+(define-constant ABSOLUTE_MINIMUM_EXIT_RC u33)
+(define-constant MAXIMUM_RC_BUFFER_FROM_PRESENT u25)
+(define-constant MINIMUM_RC_BUFFER_FROM_PRESENT u6)
+
+;; Data vars
+(define-data-var pox-reward-cycle-length uint POX_REWARD_CYCLE_LENGTH)
+(define-data-var first-burnchain-block-height uint FIRST_BURNCHAIN_BLOCK_HEIGHT)
+(define-data-var absolute-minimum-exit-rc uint ABSOLUTE_MINIMUM_EXIT_RC)
+(define-data-var configured bool false)
+
+;; This function can only be called once, when it boots up
+(define-public (set-burnchain-parameters (first-burn-height uint) (reward-cycle-length uint) (min-exit-rc uint))
+    (begin
+        (asserts! (not (var-get configured)) (err ERR_NOT_ALLOWED))
+        (var-set first-burnchain-block-height first-burn-height)
+        (var-set pox-reward-cycle-length reward-cycle-length)
+        (var-set absolute-minimum-exit-rc min-exit-rc)
+        (var-set configured true)
+        (ok true))
+)
+
+;; Stores number of votes for exit proposals by coupling it with the reward cycle the vote is cast in.
+(define-map rc-proposal-votes
+    {
+        proposed-rc: uint,
+        curr-rc: uint
+    }
+    { votes: uint }
+)
+
+;; Stores number of vetos for exit proposals by coupling it with the reward cycle the veto is cast in.
+(define-map rc-proposal-vetoes
+    {
+        proposed-rc: uint,
+        curr-rc: uint
+    }
+    { vetoes: uint }
+)
+
+;; Keeps track of miner vetos; used to ensure that a miner cannot cast multiple vetos in the same block.
+(define-map exercised-veto
+    {
+        proposed-rc: uint,
+        veto-height: uint
+    }
+    { vetoed: bool }
+)
+
+;; Keeps track of voter specific information; used to ensure a voter doesn't vote twice in the same stacking period.
+(define-map voter-state
+    { address: principal }
```

Comment:
If `voter-state`'s map key is a tuple with just one type, you can just omit the tuple and use that inner type directly:

```
(define-map voter-state
   principal
   { 
      ;; value fields
    }
 )
 ```